File: Creational/Lab3/C#/RealEstateCreator.cs
namespace C_;

public abstract class RealEstateCreator
{
    public abstract IRentableAndSellable CreateEstate();

    public void RentEstate(Customer customer, DateTime startDate, DateTime endDate){
        var estate = CreateEstate();
        estate.Rent(customer, startDate, endDate);
    }

    public void SellEstate(Customer customer){
        var estate = CreateEstate();
        estate.Sell(customer);
    }
}

File: Creational/Lab3/Python/real_estate_creator.py
from abc import *

class RealEstateCreator(ABC):
    @abstractmethod
    def create_estate(self):
        pass

    def rent_estate(self, customer, start_date, end_date):
        estate = self.create_estate()
        estate.rent(customer, start_date, end_date)

    def sell_estate(self, customer):
        estate = self.create_estate()
        estate.sell(customer)